File: EducationalScripts/Example2.rb
java_import org.apache.commons.io.IOUtils
java_import java.nio.charset.StandardCharsets
java_import org.apache.nifi.processor.io.StreamCallback

additionalContent = "test"
$flowFileAttr = {}
newAttrs = { "newAttr1" => "apple",
             "newAttr2" => "12" }

class ModifyContent
  include Java::OrgApacheNifiProcessorIo::StreamCallback

  def process(inputStream, outputStream)
    flowFileContent = Java::OrgApacheCommonsIo::IOUtils.toString(inputStream, Java::JavaNioCharset::StandardCharsets::UTF_8)
    $flowFileAttr.each do |key, value|
      flowFileContent += "\n" + (key).to_s + ": " +  (value).to_s
    end
    outputStream.write(flowFileContent.to_java_bytes)
  end
end

flowFile = session.get()

if flowFile
  begin
    $flowFileAttr = flowFile.getAttributes()
    session.write(flowFile, ModifyContent.new)
    log.info("Content has been modified!")
    flowFile = session.putAttribute(flowFile, "modify", "true")
    flowFile = session.putAllAttributes(flowFile, newAttrs)

  rescue Exception => e
    flowFile = session.putAttribute(flowFile, "error", e.to_s)
    session.transfer(flowFile, REL_FAILURE)
  else
    session.transfer(flowFile, REL_SUCCESS)
  ensure
    session.commit()
  end
end

File: EducationalScripts/Example2.js
var StreamCallback = Java.type("org.apache.nifi.processor.io.StreamCallback");
var IOUtils = Java.type("org.apache.commons.io.IOUtils");
var StandardCharsets = Java.type("java.nio.charset.StandardCharsets");


var additionalContent = "test";
var flowFileAttr = {};
var newAttrs = { "newAttr1": "apple",
                "newAttr2": String(12)};

var flowFile = session.get();
if (flowFile != null){
    try {
        flowFileAttr = flowFile.getAttributes()
        flowFile = session.write(flowFile,
            new StreamCallback(function(inputStream, outputStream) {
                var flowFileContent = IOUtils.toString(inputStream, StandardCharsets.UTF_8)
                for (var key in flowFileAttr) {
                    flowFileContent += "\n" + String(key) + ": " + String(flowFileAttr[key])
                }
                outputStream.write(flowFileContent.getBytes(StandardCharsets.UTF_8));
            }))
        log.info(String('Content has been modify!'));
        flowFile = session.putAttribute(flowFile, "modify", "true");
        flowFile = session.putAllAttributes(flowFile, newAttrs);
        session.transfer(flowFile, REL_SUCCESS);
    } catch (error) {
        flowFile = session.putAttribute(flowFile, "error", String(error));
        session.transfer(flowFile, REL_FAILURE);
    }
    session.commit()
}
